rfc-drafts/stream.md

```diff
@@ -0,0 +1,186 @@
+    
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, None)
+    }
+}
+```
+
+The arguments to `poll_next` match that of the [`Future::poll`] method:
+
+* The self must be a pinned reference, ensuring both unique access to
+  the stream and that the stream value itself will not move. Pinning
+  allows the stream to save pointers into itself when it suspends,
+  which will be required to support generator syntax at some point.
+* The [context] `cx` defines details of the current task. In particular,
+  it gives access to the [`Waker`] for the task, which will allow the
+  task to be re-awoken once data is ready.
+
+[`Future::poll`]: https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll
+[pinned]: https://doc.rust-lang.org/std/pin/struct.Pin.html
+[context]: https://doc.rust-lang.org/std/task/struct.Context.html
+[Waker]: https://doc.rust-lang.org/std/task/struct.Waker.html
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This section goes into details about various aspects of the design and
+why they ended up the way they did.
+
+## Why use a `poll` method?
+
+An alternative design for the stream trait would be to have a trait
+that defines an async `next` method:
+
+```rust
+trait Stream {
+    type Item;
+    
+    async fn next(&mut self) -> Option<Self::Item>;
+}
+```
+
+Unfortunately, async methods in traits are not currently supported,
+and there [are a number of challenges to be
+resolved](https://rust-lang.github.io/wg-async-foundations/design_notes/async_fn_in_traits.html)
+before they can be added. 
+
+Moreover, it is not clear yet how to make traits that contain async
+functions be `dyn` safe, and it is imporant to be able to pass around `dyn
+Stream` values without the need to monomorphize the functions that work
+with them.
+
+Unfortunately, the use of poll does mean that it is harder to write
+stream implementations.
+
+## What about combinators?
+
+The `Iterator` trait defines a number of useful combinators, like
+`map`.  The `Stream` trait being proposed here does not include any
+such conveniences.  Instead, they are available via extension traits,
+such as the [`StreamExt`] trait offered by the [`futures`] crate.
+
+[`StreamExt`]: https://docs.rs/futures/0.3.5/futures/stream/trait.StreamExt.html
+[`futures`]: https://crates.io/crates/futures
+
+The reason that we have chosen to exclude combinators is that a number
+of them would require access to async closures. As of this writing,
+async closures are unstable and there are a number of [outstanding
+design issues] to be resolved before they are added. Therefore, we've
+decided to enable progress on the stream trait by stabilizing a core,
+and to come back to the problem of extending it with combinators.
+
+[outstanding design issues]: https://rust-lang.github.io/wg-async-foundations/design_notes/async_closures.html
+
+This path does carry some risk. Adding combinator methods can cause
```

Comment:
Assuming the futures crate is going to start reexporting std's Stream type when built against a suitable rust version, the ambiguity issue is going to exist regardless of when we add helper methods.

Reply:
I'm not sure I follow. You're saying that if the futures crate does a 

```rust
pub use std::stream::Stream
```

or whatever, then this will trigger ambiguity?

Obviously it all depends on what mechanism we imagine. The most obvious is something like tagging the method as `#[low_priority]`, in which case we will prefer extension methods to the inherent method. It doesn't seem like it would matter then when we make the change.

Thinking more bout what I wrote, though, I was imagining that maybe we could do something like tag methods as `#[edition(2021)]` such that, in code before the edition, they are not visible, but I'm not sure what we would do *after* the edition. I guess that we could rewrite code to use `<Foo as StreamExt>::flat_map` or whatever but...nobody wants that.

Another option might be to *tag the `StreamExt` methods* so that they explicitly "cede" priority to std -- i.e., if we work in collaboration with itertools, futures, etc then people can seamlessly migrate to std.

Reply:
Oh sorry yeah, I was talking about risk in the absence of a new mechanism like `#[low_priority]`.

Reply:
OK -- but even without that mechanism, if we never move methods over, we don't have a problem, right? It's only a problem once we start to move methods from the extension trait into the regular trait.

Reply:
Ah, but we *will* have problems if we move some methods over now.

Reply:
(But presumably we would remove them from `StreamExt` as part of the minor update.)

Reply:
Would code that invokes StreamExt methods via UFCS still work if the methods are moved to a super-trait?

Reply:
I think so, but it's worth testing

Reply:
Yes [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=916f8bea3af6af3c7f75c28fdbcc3f37)